Feat(eos_designs): Custom interface descriptions

gusmb opened the pull request:
## Change Summary

Support custom descriptions for:

- Underlay ethernet interfaces (L3 and L2)
- Underlay Port-Channel interfaces
- Connected endpoints interfaces (L3 and L2)
- Tenant SVIs
- Overlay Loopback description
- VTEP loopback description

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [x] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #1131

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
In many cases, custom interface description are required to integrate with in-house monitoring systems following special patterns. This PR covers interface description customization at different levels described below in detail

### P2P and Port-Channel uplinks

P2P L3 and L2 uplinks can be customized by providing custom templates with custom logic to build the interface descriptions. Contextual facts such as `link.*` can be leveraged to facilitate the interface description calculation. Interface description default templates are defined in `defaults/default-templates.yml` and can be overwritten per switch type by providing custom template paths for each node-type:

```yaml
# Node type definitions
node_type_keys:
  spine:
    type: spine
    default_evpn_role: server
    interface_descriptions:
      # Override interface description templates with our custom templates
      underlay_ethernet_interfaces: 'interface-descriptions/underlay/ethernet-interfaces.j2'
      underlay_port_channel_interfaces: 'interface-descriptions/underlay/port-channel-interfaces.j2'
      connected_endpoints_ethernet_interfaces: 'interface-descriptions/connected-endpoints/ethernet-interfaces.j2'
      connected_endpoints_port_channel_interfaces: 'interface-descriptions/connected-endpoints/port-channel-interfaces.j2'
```

This information is included as part of the switch facts (switch.interface_descriptions):

```
{# switch.interface_descriptions #}
  interface_descriptions: {{ templates.interface_descriptions | combine(ns.node_type_key_data.interface_descriptions | arista.avd.default({})) }}
```

This fact is then used by the corresponding templates for underlay interfaces to set the interface description. Example:

```
description: "{% include switch.interface_descriptions.underlay_ethernet_interfaces %}"
```

Custom interface descriptions for P2P uplinks can also leverage `structured_config` to introduce custom data relevant to the insterface descrption, like a list of monitoring tags, or a single tag:

```yaml
... [omitted]
node_groups:
    DC1_LEAF1:
      platform: 7050SX3
      filter:
        tenants: [ all ]
        tags: [ web, app ]
      nodes:
        DC1-LEAF1A:
          id: 1
          rack: "rackA"
          mgmt_ip: 192.168.200.105/24
          mac_address: '0c:1d:c0:1d:62:01'
          uplink_switch_interfaces: [ Ethernet1, Ethernet1, Ethernet1, Ethernet1 ]
          vtep_loopback: Loopback0
          structured_config:
            tags: no-monitor
```

The `structured_config` is then passed to the switch facts (switch.struct_cfg) and can be used in a custom template:

```
{%- if link.type is arista.avd.defined('underlay_p2p') -%}
CUSTOM_P2P_LINK_TO_{{ link.peer | upper }}_{{ link.peer_interface }} {{ switch.struct_cfg.tags | arista.avd.default() }}
{%- elif link.type is arista.avd.defined('underlay_l2') -%}
CUSTOM_{{ link.peer | upper }}_{{ link.peer_interface }} {{ switch.struct_cfg.tags | arista.avd.default() }}
{%- endif -%}
```

This would produce (Example relevant LEAF switch):

```
interface Ethernet1
   description CUSTOM_P2P_LINK_TO_DC1-SPINE1_Ethernet1 no-monitor
   no shutdown
   speed forced 100gfull
   mtu 1500
   no switchport
   ip address 172.31.255.1/31
!
interface Ethernet2
   description CUSTOM_P2P_LINK_TO_DC1-SPINE2_Ethernet1 no-monitor
   no shutdown
   speed forced 100gfull
   mtu 1500
   no switchport
   ip address 172.31.255.3/31
!
interface Ethernet3
   description CUSTOM_P2P_LINK_TO_DC1-SPINE3_Ethernet1 no-monitor
   no shutdown
   speed forced 100gfull
   mtu 1500
   no switchport
   ip address 172.31.255.5/31
!
interface Ethernet4
   description CUSTOM_P2P_LINK_TO_DC1-SPINE4_Ethernet1 no-monitor
   no shutdown
   speed forced 100gfull
   mtu 1500
   no switchport
   ip address 172.31.255.7/31
!
```

### Overlay and VTEP Loopback Interfaces
The Overlay and VTEP loopback descriptions are also customizable through template, with default templates defined in `defaults/main/default-templates.yml`:

```yaml
interface_descriptions:
      underlay_ethernet_interfaces: 'designs/l3ls-evpn/interface-descriptions/underlay/ethernet-interfaces.j2'
      underlay_port_channel_interfaces: 'designs/l3ls-evpn/interface-descriptions/underlay/port-channel-interfaces.j2'
      connected_endpoints_ethernet_interfaces: 'designs/l3ls-evpn/interface-descriptions/connected-endpoints/ethernet-interfaces.j2'
      connected_endpoints_port_channel_interfaces: 'designs/l3ls-evpn/interface-descriptions/connected-endpoints/port-channel-interfaces.j2'
      overlay_loopback_interface: 'designs/l3ls-evpn/interface-descriptions/loopback-interfaces/overlay-loopback.j2'
      vtep_loopback_interface: 'designs/l3ls-evpn/interface-descriptions/loopback-interfaces/vtep-loopback.j2'
```

The default templates can be overwritten per node type as shown below:

```yaml
node_type_keys:
  spine:
    type: spine
    default_evpn_role: server
    interface_descriptions:
      # Override interface description templates with our custom templates
      underlay_ethernet_interfaces: 'interface-descriptions/underlay/ethernet-interfaces.j2'
      underlay_port_channel_interfaces: 'interface-descriptions/underlay/port-channel-interfaces.j2'
      connected_endpoints_ethernet_interfaces: 'interface-descriptions/connected-endpoints/ethernet-interfaces.j2'
      connected_endpoints_port_channel_interfaces: 'interface-descriptions/connected-endpoints/port-channel-interfaces.j2'
  l3leaf:
    type: l3leaf
    connected_endpoints: true
    default_evpn_role: client
    mlag_support: true
    network_services:
      l2: true
      l3: true
    vtep: true
    interface_descriptions:
      # Override interface description templates with our custom templates
      underlay_ethernet_interfaces: 'interface-descriptions/underlay/ethernet-interfaces.j2'
      underlay_port_channel_interfaces: 'interface-descriptions/underlay/port-channel-interfaces.j2'
      connected_endpoints_ethernet_interfaces: 'interface-descriptions/connected-endpoints/ethernet-interfaces.j2'
      connected_endpoints_port_channel_interfaces: 'interface-descriptions/connected-endpoints/port-channel-interfaces.j2'
      overlay_loopback_interface: 'interface-descriptions/loopbacks/overlay-loopback.j2'
      vtep_loopback_interface: 'interface-descriptions/loopbacks/vtep-loopback.j2'
  l2leaf:
    type: l2leaf
    connected_endpoints: true
    mlag_support: true
    network_services:
      l2: true
    underlay_router: false
    uplink_type: port-channel
    interface_descriptions:
      # Override interface description templates with our custom templates
      underlay_ethernet_interfaces: 'interface-descriptions/underlay/ethernet-interfaces.j2'
      underlay_port_channel_interfaces: 'interface-descriptions/underlay/port-channel-interfaces.j2'
      connected_endpoints_ethernet_interfaces: 'interface-descriptions/connected-endpoints/ethernet-interfaces.j2'
      connected_endpoints_port_channel_interfaces: 'interface-descriptions/connected-endpoints/port-channel-interfaces.j2'
```

The custom templates are then combined into the switch facts and used to build the corresponding loopback description:

```
loopback_interfaces:
{% if switch.router_id is arista.avd.defined %}
{# BGP Loopback #}
  Loopback0:
    description: "{% include switch.interface_descriptions.overlay_loopback_interface %}"
    shutdown: false
    ip_address: {{ switch.router_id }}/32
{%     if switch.underlay_routing_protocol == "ospf" %}
    ospf_area: {{ underlay_ospf_area }}
{%     endif %}
{%     if switch.underlay_routing_protocol == "isis" %}
    isis_enable: EVPN_UNDERLAY
    isis_passive: true
{%     endif %}
{% endif %}
{% if switch.vtep is arista.avd.defined(true) and switch.vtep_loopback | lower != 'loopback0' %}
{# Leaf VTEP vxlan tunnel loopback source interface #}
  {{ switch.vtep_loopback }}:
    description: "{% include switch.interface_descriptions.vtep_loopback_interface %}"
    shutdown: false
    ip_address: {{ switch.vtep_ip }}/32
{%     if vtep_vvtep_ip is arista.avd.defined and switch.evpn_services_l2_only is arista.avd.defined(false) %}
    ip_address_secondaries: [{{ vtep_vvtep_ip }}]
{%     endif %}
{%     if switch.underlay_routing_protocol == "ospf" %}
    ospf_area: {{ underlay_ospf_area }}
{%     endif %}
{%     if switch.underlay_routing_protocol == "isis" %}
    isis_enable: EVPN_UNDERLAY
    isis_passive: true
{%     endif %}
{% endif %}

```

### Connected Endpoints 

Ethernet and Port-Channel interfaces towards connected endpoints can be also customized the same way as for the P2P uplinks. Default templates for connected endpoints can be overwritten per node type as showed above. The new template overwrites the default one on switch facts and is used to set the interface description on the relevant templates for connected endpoints. Example:

```
{%  set peer = connected_endpoint %}
    peer_interface: {{ endpoint_ports[loop.index0] }}
{%  set peer_interface = endpoint_ports[loop.index0] %}
    peer_type: {{ peer_type }}
    description: "{% include switch.interface_descriptions.connected_endpoints_ethernet_interfaces %}"
```

Note that `peer` and `peer_interface` are set as context variables to make them available to the custom template, to help build the new interface description format, since these are commonly used parameters.

### Tenant SVIs

This PR introduces custom SVI descriptions by adding a "description" field in the model for Tenant SVIs:

```yaml
{%             set svi_description = svi_config.description | arista.avd.default(
                              svi_name) %}
```

The interface descriptions defaults to the VLAN name as it was in the previous design. However it is now possible to specify a specific SVI description for a specific tenant:

```yaml
tenants:
  # Tenant_A Specific Information - VRFs / VLANs
  Tenant_A:
    mac_vrf_vni_base: 10000
    vrfs:
      Tenant_A_OP_Zone:
        vrf_vni: 10
        vtep_diagnostic:
          loopback: 100
          loopback_ip_range: 10.255.1.0/24
          loopback_description: CUSTOM_VTEP_DIAGNOSTICS_LOOPBACK_DESC
        svis:
          # SVI as string
          '110':
            name: Tenant_A_OP_Zone_1
            description: SVI 110 CUSTOM DESCRIPTION
            tags: ['opzone']
            enabled: True
            ip_address_virtual: 10.1.10.1/24
          # SVI as integer
          111:
            vni_override: 50111
            name: Tenant_A_OP_Zone_2
            description: SVI 111 CUSTOM DESCRIPTION
            tags: ['opzone']
            enabled: True
            ip_address_virtual: 10.1.11.1/24
            ip_helpers:
              1.1.1.1:
                source_interface: lo100
                source_vrf: MGMT
```

### Tenant VTEP diagnostics Loopback

Added a "loopback_description" field in the model to provide custom value for the loopback description, defaulting to the value that was previously in place.

Data model example:

```yaml
tenants:
  # Tenant_A Specific Information - VRFs / VLANs
  Tenant_A:
    mac_vrf_vni_base: 10000
    vrfs:
      Tenant_A_OP_Zone:
        vrf_vni: 10
        vtep_diagnostic:
          loopback: 100
          loopback_ip_range: 10.255.1.0/24
          loopback_description: CUSTOM_VTEP_DIAGNOSTICS_LOOPBACK_DESC
```

Jinja template example:

```
{%  set loopback_description = tenants[tenant].vrfs[vrf].vtep_diagnostic.loopback_description | arista.avd.default(vrf + '_VTEP_DIAGNOSTICS') %}
    description: {{ loopback_description }}
```

## How to test
The test details and test results can be checked on this PR files for the molecule tests. I have introduced custom interface descriptions for all use cases as part of the test scenario.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

gusmb committed:
Basic changes to design facts adding support for custom templates for uplink interface descriptions

gusmb committed:
Additional changes. Injecting variables to interface description templates

gusmb committed:
Including vars for p2p l3 inteface description template

gusmb committed:
update

gusmb committed:
update to Jinja templates

gusmb committed:
update to template naming

gusmb committed:
update to template naming

gusmb committed:
update to jinja formatting

gusmb committed:
Added custom interface description templates

gusmb committed:
added new templates

gusmb committed:
custom loopback interface descripitons

gusmb committed:
new custom interface descriptions for tenants

gusmb committed:
testing changes

gusmb committed:
testing custom templates

gusmb committed:
update to structure config  on LEAF1A node

gusmb committed:
updates to doc

gusmb committed:
update to tests

gusmb committed:
update to doc

gusmb committed:
update doc

gusmb committed:
update doc

gusmb committed:
Removing unneded property

gusmb commented:
PR is ready for review. I had to push an additional commit to clean up unneeded data from previous commits

carlbuchmann commented:
@gusmb Thank you for this PR! This looks great. I will first review the proposed data model with my peers.

titom73 commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenants/vlan-interfaces.j2:
```diff
@@ -13,6 +13,8 @@ vlan_interfaces:
 {%             set svi_name = svi_config.name | arista.avd.default(
                               svi_profile.name) %}
+{%             set svi_description = svi_config.description | arista.avd.default(
+                              svi_name) %}
```
Missing extra whitespaces

gusmb replied:
I can't see where the extra spaces go. Lines 16 and 17 seem to be correctly aligned with 14 and 15?

gusmb replied:
Fixed extra white spaces :)

titom73 commented on ansible_collections/arista/avd/molecule/evpn_underlay_ebgp_overlay_ebgp/intended/configs/DC1-BL1A.cfg:
```diff
@@ -105,31 +105,31 @@ interface Ethernet10.200
 !
 interface Ethernet41
-   description P2P_LINK_TO_DC1-SPINE1_Ethernet6
+   description CUSTOM_P2P_LINK_TO_DC1-SPINE1_Ethernet6 
```
Seems template adds extra whitespace at EOL

gusmb replied:
True, although that is just a basic example of a customer provided template. It adds a trailing space when switch.struct_cfg.tags is empty. I will elaborate the template a bit more so it doesn't add that white space

titom73 reviewed (changes requested):
Looks really great ! Just few adjustments but LGTM overall

titom73 commented:
Suggestion: Could we make the workflow unique and move Loopback vars from main into template to be consistent across all interfaces ?

gusmb commented:
> Suggestion: Could we make the workflow unique and move Loopback vars from main into template to be consistent across all interfaces ?

I was also thinking about that and started to move it into a template, but then realized that the loopbacks are single interfaces that most likely have a static default value. Same for specific tenant svi's... 
However I have converted it into template as suggested and updated documentation. Please have a look and let me know if it makes more sense like this (updated MR description accordingly)

gusmb committed:
Remove white space EOL

gusmb committed:
Fix extra white spaces

gusmb committed:
Converted Loopback descriptions into template

gusmb committed:
remove unneeded var

titom73 reviewed (approved):
Thanks for your update on this PR! It will be very useful for lots of deployments.

LGTM!

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenants/loopback-interfaces.j2:
```diff
@@ -4,7 +4,8 @@ loopback_interfaces:
 {%         if tenants[tenant].vrfs[vrf].vtep_diagnostic is defined %}
   Loopback{{ tenants[tenant].vrfs[vrf].vtep_diagnostic.loopback }}:
-    description: {{ vrf }}_VTEP_DIAGNOSTICS
+{%  set loopback_description = tenants[tenant].vrfs[vrf].vtep_diagnostic.loopback_description | arista.avd.default(vrf + '_VTEP_DIAGNOSTICS') %}
```
Please ensure correct indenation inside the `{% %}` brackets. It would be 13 spaces in this case. Please use `~` for string concatenation instead of `+`

ClausHolbechArista replied:
We could also consider moving this into a template for consistency.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/connected_endpoints/port-channel-interfaces.j2:
```diff
@@ -67,7 +67,8 @@ port_channel_interfaces:
 {%                         set channel_group_id = (adapter.switch_ports[0] | regex_findall("\d") | join) %}
   Port-Channel{{ channel_group_id }}:
-    description: {{ connected_endpoint }}_{{ adapter_port_channel_description }}
+{%  set peer = connected_endpoint %}
```
Please fix indentation.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/extending-avd/node-types.md:
```diff
@@ -48,8 +48,45 @@ node_type_keys:
       vtep_ip_mlag: <path to J2 template - default inherited from templates.ip_addressing.vtep_ip_mlag >
       vtep_ip: <path to J2 template - default inherited from templates.ip_addressing.vtep_ip >
+
+    # Optional | Override ip_adressing templates
```
typo, "Override interface_descriptions templates"

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md:
```diff
@@ -131,8 +131,11 @@ tenants:
+            # vlan name | Required
+            name: < name >
+
+            # vlan description. | Optional
```
This should be `# SVI description | Optional`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/facts/switch/switch.j2:
```diff
@@ -117,6 +117,9 @@ switch:
 {%     set loopback_ipv4_offset = switch_data.combined.loopback_ipv4_offset | arista.avd.default(0) %}
   loopback_ipv4_offset: {{ loopback_ipv4_offset }}
 
+{# switch.loopback_ipv4_description #}
```
Please clean up after moving to template. Same for `vtep_loopback_description`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/defaults/main/main.yml:
```diff
@@ -52,6 +52,9 @@ evpn_ebgp_multihop: 3
 evpn_short_esi_prefix: "0000:0000:"
 
+# Loopback defaults
+vtep_loopback: Loopback1
```
I don't think we should combine this into the same PR unless strictly needed.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/interface-descriptions/underlay/ethernet-interfaces.j2:
```diff
@@ -0,0 +1,5 @@
+{%- if link.type is arista.avd.defined('underlay_p2p') -%}
```
I don't think we need the trim markers `{%- -%}` as they are default here.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/interface-descriptions/underlay/ethernet-interfaces.j2:
```diff
@@ -0,0 +1,5 @@
+P2P_LINK_TO_{{ link.peer | upper }}_{{ link.peer_interface }}
+{%- elif link.type is arista.avd.defined('underlay_l2') -%}
+{{ link.peer | upper }}_{{ link.peer_interface }}
+{%- endif -%}
```
I would prefer a newline at the end of the files, if it doesn't break the includes.

ClausHolbechArista reviewed (changes requested):
Nice work. I have a few comments, and please separate the suggestion of adding the global `vtep_interface` variable into it's own PR. Thanks.
Especially the nice documentation for the new templates will serve as inspiration for also improving the `ip_addressing templates` section.
Your includes are wrapped in quotes, which helps you with some of the newline behavior I have been struggling with on IP addressing. Will it work the same if you have a final newline in the included template?

ClausHolbechArista committed:
Fix review comments

ClausHolbechArista committed:
Fix review comments

titom73 committed:
Doc(eos_designs): Update interface description under node-types

ClausHolbechArista committed:
Update node-types.md
